% Generated by roxygen2 (4.1.1): do not edit by hand
% Please edit documentation in R/fetchModuGrinnNetwork.R
\name{fetchModuGrinnNetwork}
\alias{fetchModuGrinnNetwork}
\title{Compute a network module correlated to a phenotypic feature and expand the network with information from grinn internal database}
\usage{
fetchModuGrinnNetwork(datNorm, datPheno, sfPower, minModuleSize, threshold, returnAs, targetTo, filterSource, organism)
}
\arguments{
\item{datNorm}{data frame containing normalized, quantified omics data e.g. expression data, metabolite intensities.
The column names of datNorm are required to use grinn ids. \code{convertToGrinnID} is provided for id conversion, see \code{\link{convertToGrinnID}}.
Columns correspond to entities e.g. genes, metabolites, and rows to samples e.g. normals, tumors. Require 'nodetype' at the first row to indicate the type of entities in each column.}

\item{datPheno}{data frame containing phenotypic data e.g. weight, age, insulin sensitivity. Columns correspond to phenotypes and rows to samples e.g. normals, tumors.}

\item{sfPower}{numerical value of soft-thresholding power for correlation network construction. It is automatically estimated using \code{pickSoftThreshold}, or it can be defined by users.}

\item{minModuleSize}{numerical value of minimum module size for module detection.}

\item{threshold}{numerical value to define the minimum value of similarity threshold, from 0 to 1, to include edges in the output.}

\item{returnAs}{string of output type. Specify the type of the returned network.
It can be one of "tab","json","cytoscape", default is "tab". "cytoscape" is the format used in Cytoscape.js}

\item{targetTo}{string of node type. It can be one of "metabolite","protein","gene","pathway". By default, the network module will expand to pathways.}

\item{filterSource}{string or list of pathway databases. The argument is required, if \code{sourceTo} or \code{targetTo = "pathway"}.
The argument value can be any of "SMPDB","KEGG","REACTOME" or combination of them e.g. list("KEGG","REACTOME").}

\item{organism}{string of species in the following format: organism = "'species'". Default is "'Homo sapiens'".}
}
\value{
list of nodes and edges. The list is with the following componens: edges and nodes. Return empty list if found nothing
}
\description{
from input omics data e.g. normalized expression data or metabolomics data, and phenotypic data it is a one step function to:

1. Identify correlation between the input omics data e.g. normalized gene expression data, and phenotypic data e.g. weight.
The network module is the output of function \code{fetchWGCNAodule}. The function wraps around important aspects of \pkg{WGCNA} including
\code{blockwiseModules}, \code{cor}, \code{corPvalueStudent}, \code{labeledHeatmap}.
These aspects automatically perform correlation network construction, module detection, and display module-phenotype correlations.
A module or the combination of modules can be selected from the heatmap of module-phenotype correlations for including in the network output, see more details in \code{\link{fetchWGCNAodule}}.

2. Expand the network module using information from Grinn internal database.
The nodes of the network module are the keywords input to query the grinn database.
Grinn internal database contains the networks of the following types that can get expanded to:
metabolite-protein, metabolite-protein-gene, metabolite-pathway, protein-gene, protein-pathway and gene-pathway, see also \code{\link{fetchGrinnNetwork}}.
}
\examples{
# Compute a correlation of metabolite module to phenotypic data and expand to a grinn network metabolite-pathway, using information from REACTOME and KEGG only
library(grinn)
data(dummy)
data(dummyPheno)
result <- fetchModuGrinnNetwork(datNorm=dummy, datPheno=dummyPheno, minModuleSize=5, threshold=0.2, returnAs="tab", targetTo="pathway", filterSource=c("REACTOME","KEGG"))
#enter module color(s) seperate by space:yellow brown purple
#library(igraph)
#plot(graph.data.frame(result$edges[,1:2], directed=FALSE))
}
\author{
Kwanjeera W \email{kwanich@ucdavis.edu}
}
\references{
Langfelder P. and Horvath S. (2008) WGCNA: an R package for weighted correlation network analysis. BMC Bioinformatics, 9:559

Dudoit S., Yang YH., Callow MJ. and Speed TP. (2002) Statistical methods for identifying differentially expressed genes in replicated cDNA microarray experiments, STATISTICA SINICA, 12:111

Langfelder P. and Horvath S. Tutorials for the WGCNA package \url{http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/index.html}
}
\seealso{
\code{\link{cor}}, \code{\link{corPvalueStudent}}, \code{\link{pickSoftThreshold}}, \code{\link{blockwiseModules}}, \code{\link{labeledHeatmap}},
\code{\link{exportNetworkToCytoscape}}, \code{\link{fetchWGCNAodule}}, \code{\link{fetchGrinnNetwork}}, \url{http://js.cytoscape.org/}
}

