<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>grinn R package</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <img src="images/logo.png" alt="" width="450px">
      <a href="https://github.com/kwanjeeraw/grinn" class="btn">View on GitHub</a>
      <a href="https://github.com/kwanjeeraw/grinn/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/kwanjeeraw/grinn/tarball/master" class="btn">Download .tar.gz</a>
    </section>
    <section class="main-content">
      <hr>
          <h2><a id="fetchCorrGrinnNetwork" class="anchor" href="#fetchCorrGrinnNetwork" aria-hidden="true"><span class="octicon octicon-link"></span></a>fetchCorrGrinnNetwork</h2>
          <p>To compute a weighted correlation network and expand the network with information from grinn internal database</p>
          <ol>
            <li><b>INPUT</b><br>
            <p>The table summarizes important arguments</p>
            <table>
              <tr><th>Argument</th><th>Value</th><th>Description</th></tr>
              <tr><td><em>datNormX</em></td><td>data frame containing normalized-omic data</td><td>Columns correspond to entities e.g. genes and rows to samples e.g. normals, tumors.
              Require 'nodetype' at the first row to indicate the type of entities in each column.</td></tr>
              <tr><td><em>datNormY</em></td><td>data frame containing normalized-omic data</td><td>It uses the same format as <em>datNormX</em>.
              Or it can be <em>NULL</em>, if there is only one omic dataset</td></tr>
              <tr><td><em>corrCoef</em></td><td>numerical value from 0-1</td><td>The minimum absolute correlations to include edges in the output</td></tr>
              <tr><td><em>pval</em></td><td>numerical value</td><td>The maximum pvalues, to include edges in the output</td></tr>                        
              <tr><td><em>sourceTo</em></td><td>entity type e.g. metabolite</td><td>The correlation network can be expand from <em>datNormX</em> entites to a specific entity type,
              by providing a value to <em>sourceTo</em></td></tr>
              <tr><td><em>targetTo</em></td><td>entity type e.g. gene</td><td>The correlation network can be expand from <em>datNormY</em> entites (if not <em>NULL</em>) to a specific entity type,
              by providing a value to <em>targetTo</em></td></tr>
            </table>
            </li><br>
            <li><b>EXECUTE FUNCTION</b><br>
              <p>Compute a correlation network of metabolites and genes and expand to the grinn network of metabolite-protein and gene-pathway</p>
              <pre><code>
#1. load metabolomics data from a csv file
datMet = read.csv("Lung_MET.csv", header=TRUE, row.names=1, stringsAsFactors=FALSE)
#2. use only a subset of metabolomics data
datMet = datMet[c(1:7,42:50),]
#3. show dimensions of metabolomics data
dim(datMet)
#4. show the first 10 rows and 10 columns
datMet[1:10,1:10]

#!!--- NOTE: The following codes convert kegg ids to grinn ids. If the input data is already the grinn ids, these steps can be skipped.
grinnID = convertToGrinnID(txtInput=colnames(datMet), nodetype="metabolite", dbXref="kegg") #call grinn function to convert ids
grinnID = grinnID[!duplicated(grinnID[,1]),] #keep the first mapped id
colnames(datMet) = lapply(colnames(datMet),function(x) ifelse(length(which(grinnID$FROM_kegg == x))>0,as.character(grinnID$GRINNID[which(grinnID$FROM_kegg == x)]),x))
#---------- END id conversion ----------#

#5. load transcriptomics data from a csv file
datGene = read.csv("Lung_GENE.csv", header=TRUE, row.names=1, stringsAsFactors=FALSE)
#6. show dimensions of phenotypic data
dim(datGene)
#7. show the first 10 rows and 10 columns
datGene[1:10,1:10]
#8. execute function
result <- fetchCorrGrinnNetwork(datNormX=datMet, datNormY=datGene, corrCoef=0.7, pval=1e-6, method="spearman", returnAs="tab", sourceTo="protein", targetTo="pathway")
#display the first 10 edgelists
result$edges[1:10,]
              </code></pre>
            </li>
            <li><b>EXPORT OUTPUT</b><br>
              <p>Export the network as tab-delimited files to visualize in <a href="http://cytoscape.org/" target="_blank">Cytoscape</a></p>
              <pre><code>
write.table(as.matrix(result$edges),"corrNwEdge.txt",sep="\t",row.names = F, quote = FALSE)
write.table(as.matrix(result$nodes),"corrNwNode.txt",sep="\t",row.names = F, quote = FALSE)
              </code></pre>
            </li>
            <li><b>VISUALIZATION</b>
              <p>The figure is generated by Cytoscape 3.1.1 using grinn style (<em>grinn.xml</em>). It is corresponding to the cytoscape file <em>corrNw.cys</em>.</p>
            </li>
            <div id="imgScale">
              <img src="images/corrNw.png"><br>
            </div>
            <p><em>Diagram legend</em></p>
            <div id="imgScale">
              <img src="images/legend.png" width="260"><br>
            </div>
          </ol>
          <h3><a id="references" class="anchor" href="#references" aria-hidden="true"><span class="octicon octicon-link"></span></a>References</h3>
          <p>Metabolomics and transcriptomics data used in this example are taken from the following publication:</p>
          <ul>
            <li>Wikoff WR, <em>et al.</em> Metabolomic markers of altered nucleotide metabolism in early stage adenocarcinoma. <a href="http://cancerpreventionresearch.aacrjournals.org/cgi/pmidlookup?view=long&pmid=25657018" target="_blank">Cancer Prev Res (Phila) 2015;8(5):410-8</a>.</li>
            <li>Zakaria N, <em>et al.</em> Human non-small cell lung cancer expresses putative cancer stem cell markers and exhibits the transcriptomic profile of multipotent cells. <a href="http://www.biomedcentral.com/1471-2407/15/84" target="_blank">BMC Cancer 2015;15:84</a>.</li>
          </ul>
          <p>Go to <a href="index.html">HOME</a> | <a href="fgrinn.html">fetchGrinnNetwork</a> | <a href="fcorrgrinn.html">fetchCorrGrinnNetwork</a> | <a href="fdiffcorr.html">fetchDiffCorrGrinnNetwork</a>
          | <a href="fmodule.html">fetchModuGrinnNetwork</a> | <a href="fgrinncorr.html">fetchGrinnCorrNetwork</a></p>
        <hr>
        <footer>
          &copy;KW UC Davis, 2015, Grinn is maintained by <a href="https://github.com/kwanjeeraw">kwanjeeraw</a><br>
          This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>
      </section>
          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-61396514-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>
  </body>
</html>
