<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>grinn R package</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <img src="images/logo.png" alt="" width="450px">
      <a href="https://github.com/kwanjeeraw/grinn" class="btn">View on GitHub</a>
      <a href="https://github.com/kwanjeeraw/grinn/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/kwanjeeraw/grinn/tarball/master" class="btn">Download .tar.gz</a>
    </section>
    <section class="main-content">
      <hr>
          <h2><a id="fetchGrinnModuNetwork" class="anchor" href="#fetchGrinnModuNetwork" aria-hidden="true"><span class="octicon octicon-link"></span></a>fetchGrinnModuNetwork - Combine a grinn network queried from Grinn internal database to a network module correlated to a phenotypic feature</h2>
          <h3>Description</h3>
          <p>from the list of keywords, input omics data e.g. normalized expression data or metabolomics data, and phenotypic data it is a one step function to:
          <br>1. Build an integrated network (grinn network) by connecting these keywords to a specified node type, see fetchGrinnNetwork. The keywords can be any of these node types: metabolite, protein, gene and pathway. Grinn internal database contains the networks of the following types that can be quried: metabolite-protein, metabolite-protein-gene, metabolite-pathway, protein-gene, protein-pathway and gene-pathway.
          <br>2. Identify correlation between the input omics data e.g. normalized gene expression data, and phenotypic data e.g. weight. The function wraps around important aspects of WGCNA including blockwiseModules, cor, corPvalueStudent, labeledHeatmap. These aspects automatically perform correlation network construction, module detection, and display module-phenotype correlations. A module or the combination of modules can be selected from the heatmap of module-phenotype correlations for including in the network output.
          <br>3. Combine the grinn network to the network module.</p>
          <h3>Usage</h3>
          <p>fetchGrinnModuNetwork(txtInput, from, to, filterSource, returnAs, dbXref, datX, datPheno, sfPower, minModuleSize, threshold)</p>
          <h3>Arguments</h3>
          <table>
            <tr><td>txtInput</td><td>list of keywords containing keyword ids e.g. txtInput = list('id1', 'id2'). The keyword ids are from the specified database, see dbXref. Default is grinn id e.g. G371.</td></tr>
            <tr><td>from</td><td>string of start node. It can be one of "metabolite","protein","gene","pathway".</td></tr>
            <tr><td>to</td><td>string of end node. It can be one of "metabolite","protein","gene","pathway".</td></tr>
            <tr><td>filterSource</td><td>string or list of pathway databases. The argument is required, if from or to = "pathway", see from and to. The argument value can be any of "SMPDB","KEGG","REACTOME" or combination of them e.g. list("KEGG","REACTOME").</td></tr>
            <tr><td>returnAs</td><td>string of output type. Specify the type of the returned network. It can be one of "tab","json","cytoscape", default is "tab". "cytoscape" is the format used in Cytoscape.js</td></tr>
            <tr><td>dbXref</td><td>string of database name. Specify the database name used for the txtInput ids, see txtInput. It can be one of "grinn","chebi","kegg","pubchem","inchi","hmdb","smpdb","reactome","uniprot","ensembl","entrezgene". Default is "grinn". If pubchem is used, it has to be pubchem SID (substance ID).</td></tr>
            <tr><td>datX</td><td>data frame containing normalized, quantified omics data e.g. expression data, metabolite intensities. Columns correspond to entities e.g. genes, metabolites, and rows to samples e.g. normals, tumors. Require 'nodetype' at the first row to indicate the type of entities in each column.</td></tr>
            <tr><td>datPheno</td><td>data frame containing phenotypic data e.g. weight, age, insulin sensitivity. Columns correspond to phenotypes and rows to samples e.g. normals, tumors.</td></tr>
            <tr><td>sfPower</td><td>numerical value of soft-thresholding power for correlation network construction. It is automatically estimated using pickSoftThreshold, or it can be defined by users.</td></tr>
            <tr><td>minModuleSize</td><td>numerical value of minimum module size for module detection.</td></tr>
            <tr><td>threshold</td><td>numerical value to define the minimum value of similarity threshold, from 0 to 1, to include edges in the output.</td></tr>
          </table>
          <h3>Value</h3>
          <p>list of nodes and edges. The list is with the following componens: edges and nodes. Return empty list if found nothing</p>
          <h3>Examples</h3>
          <pre><code>
# Create metabolite-gene network from the list of metabolites using grinn ids and combine the grinn network to a correlation of metabolite module to phenotypic data
kw <- c('G160','G300','G371','G16414','G17191')
library(grinn)
data(dummy)
data(dummyPheno)
result <- fetchGrinnModuNetwork(txtInput=kw, from="metabolite", to="gene", datX=dummy, datPheno=dummyPheno, minModuleSize=5, threshold=0.2)
# enter module color(s) seperate by space:yellow brown purple
library(igraph)
plot(graph.data.frame(result$edges[,1:2], directed=FALSE))
          </pre></code>
          <h3><a id="references" class="anchor" href="#references" aria-hidden="true"><span class="octicon octicon-link"></span></a>References</h3>
          <p>Correlation-based analyses apply methods from the following publications:</p>
          <ul>
            <li>Dudoit S, <em>et al.</em> Statistical methods for identifying differentially expressed genes in replicated cDNA microarray experiments, <a href="http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.117.9702" target="_blank">STATISTICA SINICA 2002;12:111</a>.</li>
            <li>Langfelder P, <em>et al.</em> WGCNA: an R package for weighted correlation network analysis. <a href="http://www.biomedcentral.com/1471-2105/9/559" target="_blank">BMC Bioinformatics 2008;9:559</a>.</li>
            <li>Langfelder P, <em>et al.</em> Tutorials for the WGCNA package, <a href="http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/index.html" target="_blank"> click here</a></li> 
          </ul>
          <p>Go to <a href="index.html">HOME</a> | <a href="document.html">Documentation</a> | <a href="fgrinn.html">fetchGrinnNetwork</a> | <a href="fcorrgrinn.html">fetchCorrGrinnNetwork</a> | <a href="fdiffcorr.html">fetchDiffCorrGrinnNetwork</a>
          | <a href="fmodule.html">fetchModuGrinnNetwork</a> | <a href="fgrinncorr.html">fetchGrinnCorrNetwork</a> | <a href="finput.html">input formats</a></p>
        <hr>
        <footer>
          &copy;KW UC Davis, 2016, Grinn is maintained by <a href="https://github.com/kwanjeeraw">kwanjeeraw</a><br>
          This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>
      </section>
          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-61396514-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>
  </body>
</html>
